[QUOTA] Quota interfaces

ankagrawal opened the pull request:
This PR introduces generic interfaces for quota based throttling in ambry.

ankagrawal commented on ambry-api/src/main/java/com/github/ambry/config/QuotaConfig.java:
```diff
@@ -0,0 +1,129 @@
+  public boolean requestQuotaThrottlingEnabled;
+
+  /**
+   * Config to enable throttling for host's hardware resources quota.
```
@justinlin-linkedin Added a comment to indicate this will enable HostLevelThrottler.

ankagrawal commented on ambry-api/src/main/java/com/github/ambry/config/QuotaConfig.java:
```diff
@@ -0,0 +1,129 @@
+   *       "source": "com.github.ambry.quota.QuotaSource"
+   *     },
+   *     {
+   *       "enforcer":"com.github.ambry.quota.AnotherQuotaEnforcer",
```
Using a different class name here since QuotaEnforcer and QuotaSource has already been used in the first item.

ankagrawal commented on ambry-api/src/main/java/com/github/ambry/quota/QuotaMetric.java:
```diff
@@ -0,0 +1,27 @@
+ * Enum representing various types of quota metrics supported. It would be used by {@link QuotaSource} to get the
+ * appropriate quota values.
+ */
+public enum QuotaMetric {
```
@justinlin-linkedin The QuotaMetric will be used by QuotaSource to get the relevant quota value. For example, for a particular Ambry container (which is a QuotaResource), there would be capacity_unit quota, storage quota etc. When we query quota value from a QuotaSource, we need to also give it the list of quota metrics we are interested in.

ankagrawal commented on ambry-api/src/main/java/com/github/ambry/quota/RequestCost.java:
```diff
@@ -0,0 +1,69 @@
+ * Class representing collection of costs associated with metrics of interest of a request for a resource.
+ */
+public class RequestCost {
+  private final Map<QuotaMetric, Double> costMap;
```
@justinlin-linkedin You had this comment on the earlier PR

What I don't get about this class is whether this is a per request or not? Given by it's name and the fields quotaResource and quotaOperation, it seems like every request would have its own RequestCost. Based on that, here is my question:

Why there is a costMap, whose key is QuotaMetric?
What is the difference between QuotaMetric and QuotaResourceType here.

ankagrawal replied:
This cost is always per request.
When a caller asks for cost of a request, the cost can be of multiple types. Like cost in terms of CAPACITY_UNIT, cost in terms of STORAGE_IN_GB etc. Thats the reason why we have a map of QuotaMetric to cost here. When the QuotaEnforcer returns the cost to the caller, it will specify what is the cost for each metric. The caller can then decide, based on this cost whether to throttle or not. For example, we might say that for delivery cluster its ok to cross the CAPACITY_UNIT quota, but its not ok to cross the STORAGE_IN_GB quota. Also by returning the cost here, we have the option of returning cost of a request to the ambry client in form of response headers.
QuotaResourceType means whether the quota is for container or account or the host level quota. Its is the type of the resource for which quota is enforced. QuotaMetric is the type of quota for the resource. For example, we can have QuotaRequestType as AmbryAccount and QuotaMetric as StorageInGb.

justinlin-linkedin commented on ambry-api/src/main/java/com/github/ambry/config/QuotaConfig.java:
```diff
@@ -0,0 +1,129 @@
+  /**
+   * {@link Set} of host quota enforcers. Specified as comma separated list of enforcer classes.
+   */
+  @Config(HOST_QUOTA_ENFORCER_FACTORIES)
```
I don't think we should separate host level throttler and request quota throttler. We can unify them with the same interface and just only one json configuration to include all of them.

ankagrawal replied:
done.

justinlin-linkedin commented on ambry-api/src/main/java/com/github/ambry/config/StorageQuotaConfig.java:
```diff
@@ -138,8 +138,8 @@ public StorageQuotaConfig(VerifiableProperties verifiableProperties) {
+        QuotaMode.valueOf(verifiableProperties.getString(ENFORCER_MODE, DEFAULT_VALUE_ENFORCE_MODE).toUpperCase());
   }
 
   private String capitalize(String str) {
```
you can remove the capitalize method.

ankagrawal replied:
done.

justinlin-linkedin commented on ambry-api/src/main/java/com/github/ambry/quota/AmbryEnforcementRecommendation.java:
```diff
@@ -0,0 +1,67 @@
+/**
+ * Copyright 2020 LinkedIn Corp. All rights reserved.
```
The year please.

ankagrawal replied:
done.

justinlin-linkedin commented on ambry-api/src/main/java/com/github/ambry/quota/EnforcementRecommendation.java:
```diff
@@ -0,0 +1,48 @@
+ * request or not, along with usage information like usage percentage, name of the enforcer that made this recommendation,
+ * the recommended http status (indicating whether or not throttled request should be retried) and cost to serve the request.
+ */
+public interface EnforcementRecommendation {
```
do we really need this one to be an interface? can we just use AmbryEnforcementRecommendation every where?

ankagrawal replied:
I want to leave it as an interface for now because different implementations might want to implement it differently.

justinlin-linkedin commented on ambry-api/src/main/java/com/github/ambry/quota/HostQuotaEnforcer.java:
```diff
@@ -0,0 +1,38 @@
+ * enforcement doesn't need details about the request and can be done much before than deserialization of the request.
+ * A {@link HostQuotaEnforcer} object would usually need a {@link QuotaSource} to get and save quota and usage.
+ */
+public interface HostQuotaEnforcer {
```
Like I commented above, can we unify this interface with RequestQuotaEnforcer.

ankagrawal replied:
done.

ankagrawal commented on ambry-api/src/main/java/com/github/ambry/config/QuotaConfig.java:
```diff
@@ -0,0 +1,129 @@
+  public static final String ENFORCER_STR = "enforcer";
+  public static final String SOURCE_STR = "source";
+  public static final String QUOTA_CONFIG_PREFIX = "quota.";
+  public static final String REQUEST_QUOTA_THROTTLING_ENABLED = QUOTA_CONFIG_PREFIX + "request.throttling.enabled";
```
done.

ankagrawal commented on ambry-api/src/main/java/com/github/ambry/config/QuotaConfig.java:
```diff
@@ -0,0 +1,129 @@
+  public static final String SOURCE_STR = "source";
+  public static final String QUOTA_CONFIG_PREFIX = "quota.";
+  public static final String REQUEST_QUOTA_THROTTLING_ENABLED = QUOTA_CONFIG_PREFIX + "request.throttling.enabled";
+  public static final String HOST_QUOTA_THROTTLING_ENABLED = QUOTA_CONFIG_PREFIX + "host.throttling.enabled";
```
done.

ankagrawal commented on ambry-api/src/main/java/com/github/ambry/config/QuotaConfig.java:
```diff
@@ -0,0 +1,129 @@
+  public static final String SOURCE_STR = "source";
+  public static final String QUOTA_CONFIG_PREFIX = "quota.";
+  public static final String REQUEST_QUOTA_THROTTLING_ENABLED = QUOTA_CONFIG_PREFIX + "request.throttling.enabled";
+  public static final String HOST_QUOTA_THROTTLING_ENABLED = QUOTA_CONFIG_PREFIX + "host.throttling.enabled";
```
Removed this code.

ankagrawal commented on ambry-api/src/main/java/com/github/ambry/config/QuotaConfig.java:
```diff
@@ -0,0 +1,129 @@
+  public static final String REQUEST_QUOTA_ENFORCER_SOURCE_PAIR_INFO_JSON =
+      QUOTA_CONFIG_PREFIX + "request.enforcer.source.pair.info.json";
+  public static final String HOST_QUOTA_ENFORCER_FACTORIES = QUOTA_CONFIG_PREFIX + "host.enforcer.factories";
+  public static final String QUOTA_MANAGER_FACTORY_CLASS = QUOTA_CONFIG_PREFIX + "manger.factory.class";
```
done.

justinlin-linkedin commented on ambry-api/src/main/java/com/github/ambry/quota/AmbryEnforcementRecommendation.java:
```diff
@@ -0,0 +1,75 @@
+/**
+ * An implementation of {@link EnforcementRecommendation} made by Ambry.
+ */
+public class AmbryEnforcementRecommendation implements EnforcementRecommendation {
```
I still think it's not necessary to have EnforcementRecommendation interface. This recommendation is not a class of "how to do" but "what info to pass". I can't think of any different implementation rather than AmbryEnforcementRecommendataion.

ankagrawal replied:
removed interface.

justinlin-linkedin commented on ambry-api/src/main/java/com/github/ambry/quota/AmbryThrottlingRecommendation.java:
```diff
@@ -0,0 +1,71 @@
+/**
+ *
+ */
+public class AmbryThrottlingRecommendation implements ThrottlingRecommendation {
```
Same here for AmbryThrottlingRecommendation.

ankagrawal replied:
removed.

justinlin-linkedin commented on ambry-api/src/main/java/com/github/ambry/quota/Quota.java:
```diff
@@ -0,0 +1,72 @@
+ * The quota for a particular Ambry resource.
+ * @param <T> the type of the quota value.
+ */
+public class Quota<T> {
```
T should only be numeric types, like long, int, or float, right? If so, please put a comment here.

ankagrawal replied:
done.

justinlin-linkedin commented on ambry-api/src/main/java/com/github/ambry/quota/Quota.java:
```diff
@@ -0,0 +1,72 @@
+  private final QuotaName quotaName;
+  private final T quotaValue;
+  private final QuotaResource quotaResource;
+  private final Set<QuotaOperation> quotaOperations;
```
what is the purpose of quota operations here? Does it mean that a quota is only for certain operations? Like in a `QuotaSource`, we can get a list of `Quota`s, and they only works for Post operation?

ankagrawal replied:
removed.

justinlin-linkedin commented on ambry-api/src/main/java/com/github/ambry/quota/QuotaMode.java:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2020 LinkedIn Corp. All rights reserved.
+ * Copyright 2021 LinkedIn Corp. All rights reserved.
```
This was created in 2020.

ankagrawal replied:
fixed.

justinlin-linkedin commented on ambry-api/src/main/java/com/github/ambry/quota/QuotaResource.java:
```diff
@@ -0,0 +1,53 @@
+   * Type of Ambry resource for which quota can be applied.
+   */
+  enum QuotaResourceType {
+    AMBRY_ACCOUNT, AMBRY_CONTAINER, AMBRY_HOST
```
do we need host here?

ankagrawal replied:
removed.

justinlin-linkedin commented on ambry-api/src/main/java/com/github/ambry/quota/QuotaSource.java:
```diff
@@ -0,0 +1,39 @@
+   * @param quotaName {@link QuotaName} object.
+   * @return
+   */
+  Quota getHostQuota(QuotaResource quotaResource, QuotaName quotaName);
```
do we need host quota here?

ankagrawal replied:
removed.

justinlin-linkedin commented on ambry-api/src/main/java/com/github/ambry/quota/ThrottlingRecommendation.java:
```diff
@@ -0,0 +1,46 @@
+
+
+public interface ThrottlingRecommendation {
+
```
same as other recommendation interfaces.

ankagrawal replied:
removed.

lightningrob commented on ambry-quota/src/main/java/com/github/ambry/quota/MaxThrottlePolicy.java:
```diff
@@ -0,0 +1,39 @@
+import java.util.Map;
+
+
+public class MaxThrottlePolicy implements ThrottlePolicy {
```
javadoc

ankagrawal replied:
fixed.

lightningrob commented on ambry-quota/src/main/java/com/github/ambry/quota/capacityunit/AmbryCapacityUnitQuotaEnforcer.java:
```diff
@@ -0,0 +1,87 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.quota.capacityunit;
```
capacityunit is a strange name for a package, not very intuitive.  Is this meant to be for request/bandwidth quotas as opposed to storage?  Consider renaming.

ankagrawal replied:
I do not want to name it as request because both capacity unit and storage quotas are basically quotas for user requests. Also bandwidth doesn't really convey the meaning appropriately.
My reason for using `capacityunit` as the package name is to convey `quota.capcacityunit` is a package that deals with capacityunit type of quota. While `quota.storage` is a package that deals with storage type of quota.

lightningrob commented on ambry-quota/src/main/java/com/github/ambry/quota/capacityunit/AmbryQuotaManager.java:
```diff
@@ -0,0 +1,143 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.quota.capacityunit;
```
Did you intend to put this class in this package?  Doesn't seem to belong.

ankagrawal replied:
fixed.

ankagrawal commented on ambry-api/src/main/java/com/github/ambry/config/QuotaConfig.java:
```diff
@@ -0,0 +1,105 @@
+  public static final String THROTTLING_MODE = QUOTA_CONFIG_PREFIX + "throttling.mode";
+  public static final String REQUEST_QUOTA_ENFORCER_SOURCE_PAIR_INFO_JSON =
+      QUOTA_CONFIG_PREFIX + "request.enforcer.source.pair.info.json";
+  public static final String QUOTA_MANAGER_FACTORY = QUOTA_CONFIG_PREFIX + "manger.factory.class";
```
done.

justinlin-linkedin commented on ambry-quota/src/main/java/com/github/ambry/quota/AmbryQuotaManager.java:
```diff
@@ -0,0 +1,141 @@
+    if (requestQuotaEnforcers.isEmpty()) {
+      return allowRecommendation;
+    }
+    boolean shouldThrottle = true;
```
this variable is not used later.

justinlin-linkedin commented on ambry-quota/src/main/java/com/github/ambry/quota/AmbryQuotaManager.java:
```diff
@@ -0,0 +1,141 @@
+    if (requestQuotaEnforcers.isEmpty()) {
+      return allowRecommendation;
+    }
+    boolean shouldThrottle = true;
```
same here

justinlin-linkedin commented on ambry-quota/src/main/java/com/github/ambry/quota/AmbryQuotaManager.java:
```diff
@@ -0,0 +1,141 @@
+      return allowRecommendation;
+    }
+    boolean shouldThrottle = true;
+    Iterator<QuotaEnforcer> quotaEnforcerIterator = requestQuotaEnforcers.iterator();
```
nit: can be simplified to
```
return throttlePolicy.recommand(requestQuotaEnforcers.stream().map(enforcer -> enforcer.recommend(restRequest)).collect(Collectors.toList()));
```

ankagrawal commented on ambry-test-utils/src/main/java/com/github/ambry/quota/QuotaTestUtils.java:
```diff
@@ -0,0 +1,66 @@
+  public static QuotaConfig createDummyQuotaConfig() {
+    Properties properties = new Properties();
+    properties.setProperty(StorageQuotaConfig.HELIX_PROPERTY_ROOT_PATH, "");
+    properties.setProperty(StorageQuotaConfig.ZK_CLIENT_CONNECT_ADDRESS, "");
```
done.

ankagrawal commented on ambry-test-utils/src/main/java/com/github/ambry/quota/QuotaTestUtils.java:
```diff
@@ -0,0 +1,66 @@
+      QuotaMode quotaMode) {
+    Properties properties = new Properties();
+    properties.setProperty(StorageQuotaConfig.HELIX_PROPERTY_ROOT_PATH, "");
+    properties.setProperty(StorageQuotaConfig.ZK_CLIENT_CONNECT_ADDRESS, "");
```
done.

zzmao commented:
Tow renaming comments: 
QuotaManager-> ThrottlingManager: it would be a centralized manager for all kind of quota management. In the future, HostLevelThrottling can be also added to it. 

EnforcementRecommendation -> QuotaRecommendation: since it's per quota.

justinlin-linkedin reviewed (approved):
LGTM

ankagrawal commented:
> Tow renaming comments:
> QuotaManager-> ThrottlingManager: it would be a centralized manager for all kind of quota management. In the future, HostLevelThrottling can be also added to it.
> 
> EnforcementRecommendation -> QuotaRecommendation: since it's per quota.

Renamed EnforcementRecommendation to QuotaRecommendation. 
As discussed offline, we decided to keep QuotaManager name as is, because it’s not just handling throttling. Its also handling other things like instantiating QuotaEnforcers, giving out quota usage and charging against the quota based on request cost.
